compiler/rustc_driver_impl/src/lib.rs

```diff
@@ -1237,6 +1238,26 @@ pub fn handle_options(early_dcx: &EarlyDiagCtxt, args: &[String]) -> Option<geto
         return None;
     }
 
+    // To avoid confusion, emit warning if no space
+    // between `-o` and arg, e.g.`-optimize`, `-out-dir`, is applied, see issue #142812
+    if let Some(name) = matches.opt_str("o")
+        && let Some(suspect) = args.iter().find(|arg| arg.starts_with("-o") && *arg != "-o")
+    {
+        let confusables = ["optimize", "o0", "o1", "o2", "o3", "ofast", "og", "os", "oz"];
+        if let Some(confusable) = check_confusables(&suspect, &confusables) {
```

Comment:
Discussion: hm, I would go for very simple heuristics given a `filename` supplied to `-o`:

1. Check filename against `c_compiler_flag_confusables = ["ptimize", "0", "1", "2", "3", "fast", "g", "s", "z"];` (exact match).
2. Check filename against names of `-C` codegen options (exact match).

If we have any match, consider it a high-value confusable we should warn on. Otherwise, don't bother.